### PR TITLE
Add `Expr` support to OpenQASM 3 exporter

### LINKED ISSUE #10225: Add support for exporting `Expr` values to OpenQASM 3
### What should we add?

This tracks the work labelled "PR 3" in [the initial classical-expression design document](https://github.com/Qiskit/RFCs/blob/master/0010-simple-classical-representations.md).

Fully depends on #10223.  Work can begin following only that, but to be properly useful, this also depends on #10224.

Brief overview:
- add support for exporting `Expr` values in the OpenQASM 3 exporter.

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This adds support for `Expr` nodes in `condition` fields and `SwitchCaseOp.target`.  The general `c_if` condition checking is lifted into an `Expr` node as part of the OQ3 export, so the form of them automatically becomes like the new-style outputs, depending on how `expr.lift_legacy_condition` chooses to raise them to `Expr` nodes.

As part of the refactor of the `Expression` subtree of the temporary AST, I removed `SubroutineCall`.  Its definition was severely out-of-date (a form from before even the initial arXiv draft), where it was statement-like instead of expression-like.  There is currently no support in Qiskit for representing these sorts of subroutines properly, and the OQ3 exporter would produce invalid OQ3 if it encountered something that it thought should be a a subroutine call.  This commit fixes that bug, correctly raising an exception that non-unitary instructions are not yet supported for export.

This commit also removes the special casing from the `CXGate` into the exporter itself; it was more convenient in the new structure of some internal components, and it's a data model we don't want to support anyway.

### Details and comments

Close #10225.  Depends on #10249 and #10358.  Further changelog in #10331.

I haven't verified the exact output formats from this against what the QSS stack supports yet, but will do early next week and make whatever tweaks are required.

### CODE REVIEW

**Comment by jakelishman** [comment]
Now rebased over main.

File: qiskit/qasm3/exporter.py [hunk 1]
```diff
@@ -591,38 +583,25 @@ def build_opaque_definition(self, instruction):
...
-                "Exporting subroutines with parameters is not yet supported by the OpenQASM 3"
-                " exporter. Received this instruction, which appears parameterized:"
-                f"\n{instruction}"
+    def build_gate_definition(self, gate):
+        """Builds a QuantumGateDefinition"""
+        if isinstance(gate, standard_gates.CXGate):
+            # CX gets super duper special treatment because it's the base of Terra's definition
+            # tree, but isn't an OQ3 built-in.  We use `isinstance` because we haven't fully
+            # fixed what the name/class distinction is (there's a test from the original OQ3
+            # exporter that tries a naming collision with 'cx').
```
Comment (jlapeyre) [inline qiskit/qasm3/exporter.py:592]: The same comment appears above. We probably don't need it twice. The comments are a bit mysterious. I'm not sure if they should be clarified, or maybe I just have to read more code to understand.

Comment (jakelishman) [reply qiskit/qasm3/exporter.py:592]: Which bits of it would you like me to clarify?  I had it in both places because I thought they were sufficiently far apart in the code that it warranted it, but I'm happy to reduce them if you think it'd be clearer.

Comment (jlapeyre) [reply qiskit/qasm3/exporter.py:592]: Since putting the comments in both places was intentional, I'm ok with leaving them both in place.

**Comment by jakelishman** [comment]
I've added the test case I was worried about now, so this PR is totally ready.

File: qiskit/qasm3/ast.py [hunk 2]
```diff
@@ -194,16 +165,95 @@ def __init__(
...
 
+class Constant(Expression, enum.Enum):
+    """A constant value defined by the QASM 3 spec."""
+
+    PI = enum.auto()
+    EULER = enum.auto()
+    TAU = enum.auto()
+
+
+class IntegerLiteral(Expression):
```
Comment (jlapeyre) [inline qiskit/qasm3/ast.py:189]: Renaming these classes with correct names is a great idea. There seem to be a number of these improvements that are unrelated to the  PR. One might argue for doing this in a separate PR. But in the interest of development efficiency, I think it's ok to include in this PR.

Comment (jakelishman) [reply qiskit/qasm3/ast.py:189]: In some senses there's not a lot of renaming going on here - most of the classes I removed weren't actually used, and I was putting in new nodes for the actual `Expr` tree.  All previous `ParameterExpression` values were being handled by the base `ast.Expression` node, which I _effectively_ renamed to `ast.StringifyAndPray` to make it clearer what was actually going on, and so I had a more sane base `ast.Expression` node to work from.

I think `IntegerLiteral` might be the only one that logically existed in the old form that still exists now but with a different name.

Fwiw, I think this AST module will be completely gone by the end of the year; once we have our custom importer, we can use the AST from that, rather than needed to maintain a parallel AST and formatter in Terra.
